### DISCUSSION TITLE
Possible to stream the entire orderbook?

### DISCUSSION THREAD

**Discussion by jclangner** [nautechsystems/nautilus_trader#663] [Q&A] [answered]
I have used the order book websocket on FTX before and it is possible to get the top 100 levels of the orderbook. Is this possible yet through Nautilus ?

**Comment by limx0** [comment, marked as answer]
Hey @jclangner, yes Nautilus will handle any arbitrary depth of order book.

**Reply by jclangner** [reply]
thanks for the response, any details how to do so?

**Reply by limx0** [reply]
Inside your trading strategy, you need to call `subscribe_order_book_deltas` or `subscribe_order_book_snapshots`. A simple example would be: 

```python

class OrderbookStrategyConfig(StrategyConfig):
    instrument_id: str

class OrderbookStrategy(Strategy):
    def __init__(self, config: OrderbookStrategyConfig):
        super().__init__(config)
        self.instrument_id = InstrumentId.from_str(config.instrument_id)

    def on_start(self):
        self.subscribe_order_book_deltas(self.instrument_id, depth=100)
 
     def on_order_book_delta(self, delta: OrderBookDelta):
        self.log.info(repr(delta), LogColor.CYAN)
        

```

And setup a live instance using this strategy, by modifying the FTX live trading example from [here](https://github.com/nautechsystems/nautilus_trader/blob/master/examples/live/ftx_market_maker.py)

**Reply by jclangner** [reply]
ah thanks man, was missing the depth param on subscribe_order_book_deltas

**Reply by limx0** [reply]
Although just looking at the FTX adapter, we don't appear to be passing through the depth param in the FTX adapter.

The FTX adapter receives the subscribe request [here](https://github.com/nautechsystems/nautilus_trader/blob/master/nautilus_trader/adapters/ftx/data.py#L175) and then sends off a subscribe to FTX via websocket [here](https://github.com/nautechsystems/nautilus_trader/blob/aaead1ee7a386fd054777a92220f4c501e461f53/nautilus_trader/adapters/ftx/websocket/client.py#L170) but we don't use the depth. It might be okay though because quickly looking at the FTX docs, it seems like the default is to subscribe to all levels?

Give it a try and let us know if you're not getting the full depth and we can investigate further.

**Reply by jclangner** [reply]
Seems like the right way to do this is to just subscribe to order book deltas which initially gives you the full orderbook and then just apply the deltas onto that orderbook to get continuous full order book updates. Im still having troubles when using subscribe_order_book_snapshots (for example if i only need an orderbook snapshot every 1 second). Using BookType.L2_MBP, i can't seem to get more than three order book levels when getting snapshots.

**Reply by limx0** [reply]
Copy @jclangner - this should work, but let us do some testing and come back to you. Thanks for the extra detail.

**Reply by cjdsellers** [reply]
Hi @jclangner 

Thanks for your interest in the project!
After a little investigation we figured out that we only print the top 3 levels by default with a `repr`, or by calling `pprint()` on the book.

```
    def __repr__(self) -> str:
        return (
            f"{type(self).__name__}\n"
            f"instrument: {self.instrument_id}\n"
            f"timestamp: {pd.Timestamp(self.ts_last)}\n\n"
            f"{self.pprint()}"
        )
```

```
cpdef str pprint(self, int num_levels=3, show="size")
```

A quick test with the following in the logs for `on_order_book`:
```
self.log.info("asks = " + str(len(order_book.asks.levels)), LogColor.CYAN)
self.log.info("bids = " + str(len(order_book.bids.levels)), LogColor.CYAN)
```

Shows this:
```
2022-06-04 06:11:13.999888659 UTC [INF] TESTER-001.VolatilityMarketMaker-000: asks = 100
2022-06-04 06:11:13.999892047 UTC [INF] TESTER-001.VolatilityMarketMaker-000: bids = 100
```

It looks like FTX have no `depth` param for their subscribe message, so we're getting the default 100 per side.

Hope this helps!
